presto-main/src/main/java/io/prestosql/operator/scalar/time/TimeToTimestampCast.java

```diff
@@ -0,0 +1,96 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.operator.scalar.time;
+
+import io.prestosql.spi.connector.ConnectorSession;
+import io.prestosql.spi.function.LiteralParameter;
+import io.prestosql.spi.function.LiteralParameters;
+import io.prestosql.spi.function.ScalarOperator;
+import io.prestosql.spi.function.SqlType;
+import io.prestosql.spi.type.LongTimestamp;
+import io.prestosql.spi.type.TimeType;
+
+import java.time.Instant;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.ZoneOffset;
+import java.util.List;
+
+import static io.prestosql.spi.function.OperatorType.CAST;
+import static io.prestosql.type.DateTimes.MICROSECONDS_PER_SECOND;
+import static io.prestosql.type.DateTimes.PICOSECONDS_PER_MICROSECOND;
+import static io.prestosql.type.DateTimes.PICOSECONDS_PER_SECOND;
+import static io.prestosql.type.DateTimes.SECONDS_PER_DAY;
+import static io.prestosql.type.DateTimes.rescale;
+import static io.prestosql.type.DateTimes.round;
+import static io.prestosql.type.DateTimes.scaleEpochMicrosToMillis;
+import static java.lang.Math.multiplyExact;
+
+@ScalarOperator(CAST)
+public final class TimeToTimestampCast
+{
+    private TimeToTimestampCast() {}
+
+    @LiteralParameters({"sourcePrecision", "targetPrecision"})
+    @SqlType("timestamp(targetPrecision)")
+    public static long castToShort(
+            @LiteralParameter("sourcePrecision") long sourcePrecision,
+            @LiteralParameter("targetPrecision") long targetPrecision,
+            ConnectorSession session,
+            @SqlType("time(sourcePrecision)") long time)
+    {
+        long epochMicros = computeEpochMicros(sourcePrecision, targetPrecision, session, time);
+        if (targetPrecision <= 3) {
+            return scaleEpochMicrosToMillis(epochMicros);
+        }
+
+        return epochMicros;
+    }
+
+    @LiteralParameters({"sourcePrecision", "targetPrecision"})
+    @SqlType("timestamp(targetPrecision)")
+    public static LongTimestamp castToLong(
+            @LiteralParameter("sourcePrecision") long sourcePrecision,
+            @LiteralParameter("targetPrecision") long targetPrecision,
+            ConnectorSession session,
+            @SqlType("time(sourcePrecision)") long time)
+    {
+        long epochMicros = computeEpochMicros(sourcePrecision, targetPrecision, session, time);
+        return new LongTimestamp(epochMicros, (int) (time % PICOSECONDS_PER_MICROSECOND));
+    }
+
+    private static long computeEpochMicros(@LiteralParameter("sourcePrecision") long sourcePrecision, @LiteralParameter("targetPrecision") long targetPrecision, ConnectorSession session, @SqlType("time(sourcePrecision)") long time)
+    {
+        long epochDay = LocalDate.ofInstant(session.getStart(), session.getTimeZoneKey().getZoneId())
```

Comment:
Since this can be invoked per row, we should avoid java.time. here.
I am OK with this as is, but a TODO/issue would be nice.